pkg/urbit/vere/pier.c

```diff
@@ -1584,64 +1584,88 @@ static u3_boot
 _pier_pill_parse(u3_noun pil)
 {
   u3_boot bot_u;
-  u3_noun pil_p, pil_q, pil_r;
-  u3_noun pro;
+  u3_noun pil_p, pil_q;
 
   c3_assert( c3y == u3du(pil) );
+  u3x_cell(pil, &pil_p, &pil_q);
 
-  if ( c3y == u3h(pil) ) {
-    u3x_trel(pil, 0, &pil_p, &pil_q);
-  }
-  else {
-    u3x_qual(pil, 0, &pil_p, &pil_q, &pil_r);
-  }
+  {
+    //  XX use faster cue
+    //
+    u3_noun pro = u3m_soft(0, u3ke_cue, u3k(pil_p));
+    u3_noun mot, tag, dat;
 
-  pro = u3m_soft(0, u3ke_cue, u3k(pil_p));
+    if (  (c3n == u3r_trel(pro, &mot, &tag, &dat))
+       || (u3_blip != mot) )
+    {
+      u3m_p("mot", u3h(pro));
+      fprintf(stderr, "boot: failed: unable to parse pill\r\n");
+      u3_king_bail();
+      exit(1);
+    }
 
-  if ( 0 != u3h(pro) ) {
-    fprintf(stderr, "boot: failed: unable to parse pill\r\n");
-    exit(1);
-  }
+    if ( c3y == u3r_sing_c("ivory", tag) ) {
+      fprintf(stderr, "boot: failed: unable to boot from ivory pill\r\n");
+      exit(1);
```

Comment:
We now `u3_king_bail` prior to the other `exit` calls. Why do we not do that here?

Reply:
Good catch!